rosette/src/Export.cc

```diff
@@ -35,39 +35,31 @@ void defaultObjectHandler( ObjectCodePB::Object * lvob, pOb ob, ObjectCodePB::Ob
 
 // This function looks up an object in the handlers table based on its type string
 // and then calls the appropriate handler to populate it.
-std::set<IdType> exportedIds;
 void populateObjectByType(pOb ob, ObjectCodePB::Object *exOb) {
 
-    exOb->set_objectid(BASE(ob)->objectId);
+    exOb->set_object_id(BASE(ob)->objectId);
 
-    auto id = exportedIds.find(BASE(ob)->objectId);
-    if (id != exportedIds.end()) {
-        // Don't export an object more than once. This effectively eliminates problems
-        // with self referential objects and loops in the object tree.
+    std::string type = BASE(ob)->typestring();
+    auto oh = handlers.find(type);
```

Comment:
Why did you remove the check for whether the object had already been exported?

Reply:
It seemed to be too generalized.  There is a use of objects that I don't fully understand wherein they take on different values. This caused some symbol values to not be exported.  I had originally implemented this due to the self-referencial nature of the TupleExpr object causing an endless loop and a crash.  I'm still struggling over the way objects mutate during the mixed compile and execute phases.

Here's an example of a litvec containing a table (that holds a tuple). The elements contain symbols that apparently have the same object id, yet contain different values.

code_block {
  object_id: 2649
  litvec {
    ob {
      type: OT_PROC
      object_id: 2637
      meta_id: 32
      parent_id: 33
      proc {
        env {
          type: OT_TBL_OBJECT
          object_id: 359
          meta_id: 1125
          parent_id: 357
          tbl_object {
            validextent {
              type: OT_FIXNUM
              object_id: 1640
              meta_id: 299
              parent_id: 300
              fixnum {
                value: 504
              }
            }
            keyvec {
              type: OT_TUPLE
              object_id: 1303
              meta_id: 1150
              parent_id: 9
              tuple {
                elements {
                  type: OT_SYMBOL
                  object_id: 1637
                  meta_id: 308
                  parent_id: 309
                  symbol {
                    name: "Timer-SBO"
                  }
                }
                elements {
                  type: OT_SYMBOL
                  object_id: 1637
                  meta_id: 308
                  parent_id: 309
                  symbol {
                    name: "RblStack-SBO"
                  }
                }
                elements {
                  type: OT_SYMBOL
                  object_id: 1637
                  meta_id: 308
                  parent_id: 309
                  symbol {
                    name: "Tuple-SBO"
                  }
                }
                elements {
                  type: OT_SYMBOL
                  object_id: 1637
                  meta_id: 308
                  parent_id: 309
                  symbol {
                    name: "RblTable-SBO"
                  }
                }
                elements {
                  type: OT_SYMBOL
                  object_id: 1637
                  meta_id: 308
                  parent_id: 309
                  symbol {
                    name: "Reader-SBO"
                  }
                }
.
.
.

This may be due to objects being copied without a new object id being created. I'm just not sure how all this works yet.

Reply:
> This may be due to objects being copied without a new object id being created.
That seems most likely to me, and was why I brought up the copy constructor when we talked earlier.

Reply:
> I had originally implemented this due to the self-referencial nature of the TupleExpr object causing an endless loop and a crash.

Out of curiosity: What are you referring to when you say self-referencial nature of the TupleExpr?